src/query/graphite/native/builtin_functions.go

```diff
@@ -1591,16 +1591,70 @@ func changed(ctx *common.Context, seriesList singlePathSpec) (ts.SeriesList, err
 	})
 }
 
-// movingMedian takes one metric or a wildcard seriesList followed by a a quoted string
-// with a length of time like '1hour' or '5min'. Graphs the median of the preceding
-// datapoints for each point on the graph. All previous datapoints are set to None at
-// the beginning of the graph.
-func movingMedian(ctx *common.Context, _ singlePathSpec, windowSize string) (*binaryContextShifter, error) {
-	interval, err := common.ParseInterval(windowSize)
+// windowPointsLength calculates the number of window points in a interval
+func windowPointsLength(series *ts.Series, interval time.Duration) int {
+	return int(interval / (time.Duration(series.MillisPerStep()) * time.Millisecond))
+}
+
+type movingImplementationFn func(window []float64, values ts.MutableValues, windowPoints int, i int)
+
+// movingMedianHelper given a slice of floats, calculates the median and assigns it into vals as index i
+func movingMedianHelper(window []float64, vals ts.MutableValues, windowPoints int, i int) {
+	nans := common.SafeSort(window)
+
+	if nans < windowPoints {
+		index := (windowPoints - nans) / 2
+		median := window[nans+index]
+		vals.SetValueAt(i, median)
+	}
+}
+
+// movingSumHelper given a slice of floats, calculates the sum and assigns it into vals as index i
+func movingSumHelper(window []float64, vals ts.MutableValues, windowPoints int, i int) {
+	sum, nans := common.SafeSum(window)
+
+	if nans < windowPoints {
+		vals.SetValueAt(i, sum)
+	}
+}
+
+// movingMaxHelper given a slice of floats, finds the max and assigns it into vals as index i
+func movingMaxHelper(window []float64, vals ts.MutableValues, windowPoints int, i int) {
+	max, nans := common.SafeMax(window)
+
+	if nans < windowPoints {
+		vals.SetValueAt(i, max)
+	}
+}
+
+// movingMinHelper given a slice of floats, finds the min and assigns it into vals as index i
+func movingMinHelper(window []float64, vals ts.MutableValues, windowPoints int, i int) {
+	min, nans := common.SafeMin(window)
+
+	if nans < windowPoints {
+		vals.SetValueAt(i, min)
+	}
+}
+
+
+
+func newMovingBinaryTransform(
+	ctx *common.Context,
+	input singlePathSpec,
+	windowSizeValue genericInterface,
+	movingFunctionName string,
+	impl movingImplementationFn,
+) (*binaryContextShifter, error) {
+	if len(input.Values) == 0 {
+		return nil, nil
```

Comment:
Can you add some explicit tests to make sure this case is handled ok? Would add a test case to `engine_test.go` and inside `TestExecute` to make sure a binary transform correctly is handled if a `nil` binary context shifter is returned (i.e. send input values of zero).

I just spent 10-20mins finding if it was or not and not 100% confident about it. I see this which implies it should be:
```go
	// context shifter ptr is nil, nothing to do here, return empty series.
	if result.IsNil() {
		return reflect.ValueOf(ts.NewSeriesList()), nil
	}
```

Reply:
OK.

So you said you want me to add "a test case to `engine_test.go` and inside `TestExecute`"

 i added the test case to `engine_test.go`

but i dont quite understand your wording on the second part. were you saying that you also wanted me to modify `TestExecute` to test whether or not a binary transform is correctly handled if a `nil` binary context shifter is returned? why would I need to test for that in two places?

Reply:
resolved offline. Rob said this test was enough.